### ISSUE TITLE
Behavior of json_unpack (document or update code?)

### ISSUE THREAD

**Issue by Andrade** [akheron/jansson#357]
When doing `json_unpack` with multiple values and the call fails, some of the values may be set, while others (valid) values are not. I think this behavior should either (1) be documented or (2) changed to make sure that if the call fails then none of the variables is set/changed.

To give it more context, assume the following unpack:

```
const char format[] = "{s?s, s?i, s?s}";
const char *s1, *s3;
int n;
json_unpack_ex(j, &error, 0, format,
		"string1", &s1,
		"string2", &n,
		"string3", &s3)
```

There is a mistake, the second item to unpack is in reality a string, but I marked it as an integer with `s?i`. The unpack will fail and in this case `s1` will be set but `s3` will still be null.

I do one of two things: either (1) validate the incoming json beforehand with `JSON_VALIDATE_ONLY`, or (2) reset/free all variables in case of failure.

---

I'm using libjansson from the repositories of Ubuntu 16.04, currently at version: *libjansson4 (= 2.7-3)*.

**Comment by Andrade** [comment]
I think I got a better understanding on how it works. If I'm understanding it correctly, memory for strings is allocated internally, but once we call `json_decref(j)` then all memory is released.

The reason I'm having issues is because I set all char arrays to NULL before calling the function, and after `json_decref` they (some) are no longer NULL even though they have been freed. So maybe the pointers are not being reset to their old values (in this case NULL) after we call `json_decref`.

In addition, if I want to release the resources related to Jansson but keep the char arrays for later use, then I'll have to copy them over to a new variable. In my previous code snippet:

```
const char format[] = "{s?s, s?i, s?s}";
const char *s1=NULL, *s3=NULL;
int n=0;
json_unpack_ex(j, &error, 0, format,
		"string1", &s1,
		"string2", &n,
		"string3", &s3)

//1: s1 is not NULL, but s3 is NULL
//2: somehow copy the string I need to keep into new_string
json_decref(j); // releases not only j but also s1 which was set
//3: s1 is not NULL, but s3 is NULL (like before, but...
//   now s1 only contains garbage: it's okay because I have a copy called new_string)
```

**Comment by akheron** [comment]
>When doing json_unpack with multiple values and the call fails, some of the values may be set, while others (valid) values are not. I think this behavior should either (1) be documented or (2) changed to make sure that if the call fails then none of the variables is set/changed.

Actually, this behavior **is** documented, although only in the comments of the last example near the end of [this chapter](http://jansson.readthedocs.io/en/2.10/apiref.html#parsing-and-validating-values):

```c
/* root is an empty JSON object */
int myint = 0, myint2 = 0, myint3 = 0;
json_unpack(root, "{s?i, s?[ii]}",
            "foo", &myint1,
            "bar", &myint2, &myint3);
/* myint1, myint2 or myint3 is no touched as "foo" and "bar" don't exist */
```

> I think I got a better understanding on how it works. If I'm understanding it correctly, memory for strings is allocated internally, but once we call json_decref(j) then all memory is released.

Your reasoning is correct, and this is [documented behavior](http://jansson.readthedocs.io/en/2.10/apiref.html#parsing-and-validating-values). From the description of the `s` format:

> `s` (string) [const char *]
> Convert a JSON string to a pointer to a null terminated UTF-8 string. The resulting string is extracted by using json_string_value() internally, so it exists as long as there are still references to the corresponding JSON string.


The best way to keep the string around is to use the `s%` format to obtain a pointer and data length, call `malloc` to allocate a buffer of the same length, and then copy the data using `memcpy`.

Another way is to unpack using the `O` format (capital o), and then use `json_string_value()` to obtain the string data. The string data is available until you decref the extracted `json_t` value. This has the downside that you don't get automatic validation by `json_unpack()`, so the value might not be a string.

**Comment by Andrade** [comment]
Thank you for taking the time to look into this issue, and also for the tip at the end.

I wrote this piece of code showing what I thought could be an issue. I'm adding it for completeness:

```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <jansson.h>

int main()
{
	const char format_in[] = "{s:s, s:s, s:s}";
	const char format_out[] = "{s?s, s?i, s?s}";
	json_error_t error;
	const char *s1 = NULL;
	int         n2 = 0;
	const char *s3 = NULL;
	char *s;
	
	json_t *j = json_pack(format_in,
			"string1", "foo",
			"string2", "str in pack, int in unpack",
			"string3", "bar");
	s = json_dumps(j, 0);
	printf("pack result=%s\n", s);
	free(s);
	
	int ret = json_unpack_ex(j, &error, 0, format_out, "string1", &s1, "string2", &n2, "string3", &s3);
	s = json_dumps(j, 0);
	printf("unpack result: s1=%s, n2=%d, s3=%s\n", s1, n2, s3);
	// return is -1 (fail) but some values are set and others are not
	printf("ret=%d\n", ret);
	if (ret) {
		fprintf(stderr, "error text: %s\n", error.text);
	}
	free(s);

	json_decref(j);
	
	printf("Is s1 NULL after releasing resources? %d\n", (s1==NULL));
	printf("Is s3 NULL after releasing resources? %d\n", (s3==NULL));

	return 0;
}

$ gcc -g -Wall -Wextra -o jansson3 jansson3.c -ljansson && ./jansson3
pack result={"string3": "bar", "string1": "foo", "string2": "str in pack, int in unpack"}
unpack result: s1=foo, n2=0, s3=(null)
ret=-1
error text: Expected integer, got string
Is s1 NULL after releasing resources? 0
Is s3 NULL after releasing resources? 1
```

- `json_unpack_ex` fails with `-1` because it tries reading "string2" as an int, but it is actually a string. Despite the failure with return `-1`, the first parameter `foo` is still set.
- Had I tried extracting a value named say "string4" which does not exist, instead of "string2", then the return value would have been `0` and both `s1` and `s3` would be set.
- After resources are released, `s1` is not set to NULL.

So basically, my doubt was why values were being set when the return value was -1 (and I wasn't aware at first that the library would release the strings and not the caller). This was the first time I used the library, but the behavior is clear to me now.

For my part, the issue is solved. Thanks!
